api/src/main/java/org/infinispan/api/common/events/KeyValueEvent.java

```diff
@@ -0,0 +1,11 @@
+package org.infinispan.api.common.events;
+
+import org.infinispan.api.common.KeyValueEntry;
+
+/**
+ *
+ * @since 13.0
+ **/
+public interface KeyValueEvent<K, V> {
```

Comment:
Should we have more info on here?

1. Pre/Post value
2. Metadata (Pre/post as well)
3. origin local
4. attached transaction
5. current state?

Reply:
I have added `isOriginLocal` (`isCurrentState` was already there). The tx is returned in `getSource()`.
Currently the entry event looks like this:
```java
interface CacheEntryEvent<K, V> {
   /**
    * @return The entry after the event
    */
   CacheEntry<K, V> newEntry();

   /**
    * @return The entry before the event
    */
   CacheEntry<K, V> previousEntry();
```
@danberindei why do you prefer to unroll those with individual methods ?

Reply:
I don't have a real preference @tristantarrant. In the invocation context we keep only one CacheEntry, so we'd have to create a wrapper for the previous entry (or maybe 2 wrappers, one for each entry). Unfortunately cluster/client listeners use the same API internally, so we'd create those wrappers for them on the primary node as well. But it's not that big a deal, as long as we only create those when we have a listener I think we're fine.